### ISSUE TITLE
kube-apiserver crash on invalid service replace request

### ISSUE THREAD

**Issue by gary-lloyd-tessella** [kubernetes/kubernetes#55327]
BUG REPORT

/kind bug

**What happened**:
I tried to submit a Service Replace (https://kubernetes.io/docs/api-reference/v1.8/#replace-155) to the kube-apiserver and this results the kube-apiserver container crashing and restarting.
PUT localhost:8001/api/v1/namespaces/<namespace>/services/<servicename>

**What you expected to happen**:
I expected the server not to crash and to get validation errors if the request was invalid.

**How to reproduce it (as minimally and precisely as possible)**:
Create a service, get the resourceVersion and clusterIP from the created resource
GET localhost:8001/api/v1/namespaces/<namespace>/services/<servicename>

Submit a replace command to
PUT localhost:8001/api/v1/namespaces/<namespace>/services/<servicename>
with body like

```
{
    "kind": "Service",
    "apiVersion": "v1",
    "metadata": {
        "name": "servicename",
        "resourceVersion": "<resource-Id>"
    },
    "spec": {
        "ports": [
            {
                "name": "portname",
                "port": 80,
                "targetPort": 80
            },
            {
                "name": "anotherport",
                "port": 8000,
                "targetPort": 8000
            }
        ],
        "selector": {
            "name": "rstudio-testrstudio"
        },
        "clusterIP": "<cluster-ip>",
        "type": "NodePort",
        "sessionAffinity": "None",
        "externalTrafficPolicy": "Cluster"
    },
    "status": {
        "loadBalancer": {}
    }
}
```
If the same request is submitted with the nodePort fields in the ports array then the request succeeds.
```
"ports": [
            {
                "name": "portname",
                "port": 80,
                "targetPort": 80,
                "nodePort": 32741
            },
            {
                "name": "anotherport",
                "port": 8000,
                "targetPort": 8000,
                "nodePort": 31431
            }
        ],
```
I get that I may be sending an invalid request without a node port (though shouldn't the merge deal with that?) but it shouldn't take down the API server.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.3", GitCommit:"0480917b552be33e2dba47386e51decb1a211df6", GitTreeState:"clean", BuildDate:"2017-05-10T23:29:08Z", GoVersion:"go1.8.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.1", GitCommit:"f38e43b221d08850172a9a4ea785a86a3ffa3b3a", GitTreeState:"clean", BuildDate:"2017-10-11T23:16:41Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration:
Bare metal cluster

- OS (e.g. from /etc/os-release): Ubuntu 16.04.3 LTS (Xenial Xerus)
- Kernel (e.g. `uname -a`): 4.4.0-98-generic
- Install tools:
- Others:

**Comment by xiangpengzhao** [comment]
@gary-lloyd-tessella thanks for reporting this! Seems like this is similar to #54898 and should be fixed by #49258.
@liggitt @freehan can you PTAL that PR to see if it's good to merge?

**Comment by liggitt** [comment]
Duplicate of https://github.com/kubernetes/kubernetes/issues/54898
